lib/workflow/stores/mongo.js

```diff
@@ -639,22 +639,39 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
         assert.uuid(data.taskId, 'data.taskId');
         assert.string(data.state, 'data.state');
 
-        var query = {
+        var queryDependency = {
             graphId: data.graphId,
-            reachable: true
+            reachable: true,
         };
-        query['dependencies.' + data.taskId] = {
+        queryDependency['dependencies.' + data.taskId] = {
             $in: [data.state, Constants.Task.States.Finished]
         };
-        var update = {
+        
+        var updateDependency = {
+            $unset: {}
+        };
+        updateDependency.$unset['dependencies.' + data.taskId] = '';
+
+        var queryAnyOf = {
+            graphId: data.graphId,
+            reachable: true,
+        };
+        queryAnyOf['dependencies.anyOf.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
+        var updateAnyOf = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
+        updateAnyOf.$unset["dependencies.anyOf"] = "";
+
         var options = {
             multi: true
         };
-
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        
+        return Promise.all([
```

Comment:
@brianparry , All of the promise.all() here is because there are multiple conditions exists in database,  I need to solve the all conditions in one Promise return.  Here, Condition 1: The task to update is under dependencies.<taskid>, and we need to remove dependencies.<taskId> Condition 2. The task to update is under dependencies.anyOf.<taskId>, and we need to remove dependencies.anyOf. In MongoDB we need two conditions and two differnet update operations to deal with this situation. 

Similar condition happens in all of the following Promise.all() conditions.